lib/raxx/stack.ex

```diff
@@ -0,0 +1,132 @@
+defmodule Raxx.Stack do
+  alias Raxx.Server
+  alias Raxx.Middleware
+  alias Raxx.Middleware.Pipeline
+
+  @behaviour Server
+
+  @enforce_keys [:pipeline, :server]
+  defstruct @enforce_keys
+
+  @moduledoc """
+  `Raxx.Stack` implements `Raxx.Server` behaviour that works with
+  `t:Raxx.Stack.t/0` as its state.
+  """
+
+  @typedoc """
+  A `t:Raxx.Stack.t/0` represents a pipeline of middlewares attached to a server.
+
+  NOTE: Don't rely on the internal structure of this type. It can be modified
+  at an arbitrary moment to improve performance or capabilities.
+  """
+  # DEBT: compare struct t() performance to a (tagged) tuple implementation
+  @opaque t :: %__MODULE__{
+            pipeline: Pipeline.t(),
+            server: Server.t()
+          }
+
+  @type server :: {__MODULE__, t()}
```

Comment:
I'm tempted to conflate `Stack.t` and `Stack.server` to a single

`@type t :: {__MODULE__, state}` and have all helper functions operate on this type.

Outside of the module itself I don't think there is a situation where anyone would use just the server state, without the module.

Reply:
I can't think of one either